File: haskell/prob151-200.hs
import Data.List
import Math

-- Problem 173

laminas lim = [y^2-x^2| x <- [1..limx] , y <- [x+2,x+4.. (succ limx)], (y^2) - (x^2) <= lim]
  where limx = quot lim 4

lamina_2 lim = [[y^2-x^2| y <- [x+2,x+4.. (succ limx)], (y^2) - (x^2) <= lim] | x <- [1..limx]]
  where limx = quot lim 4

sol_174 lim = sum $ map length $ lamina_2 lim

laminas_2 x y res lim 
  | tiles > lim = if x > (quot lim 4) then res else laminas_2 (succ x) (x+3) res lim
  | otherwise = laminas_2 x (y + 2) (succ res) lim
  where tiles = (y^2) - (x^2)

sol_174a lim = laminas_2 1 3 0 lim

distinct lst =  map (head.group.sort) lst

lamina_3 lim = [[d^2 + 2*x*d | d <- [2,4..limx], d^2 + 2*x*d <= lim] | x <- [1..limx]]
  where limx = quot lim 4

laminas_4 x d res lim 
  | tiles > lim = if x > (quot lim 4) then res else laminas_4 (succ x) 2 res lim
  | otherwise = laminas_2 x (d + 2) (succ res) lim
  where tiles = d^2 + 2*x*d

sol_174b lim = laminas_2 1 2 0 lim

File: clojure/prob151-200.clj
(ns euler.prob151-200)

;; PROBLEM 173/174

(load-file "math.clj")

(defn sqr [x] (* x x))

(defn add2 [x] (+ 2 x))
(defn min2 [x] (- x 2))

(defn laminas
  [n t]
  (->> (- itmp jtmp)
       (for [j (range (min2 i) 0 -2)
             :let [jtmp (sqr j)]
             :while (<= (- itmp jtmp) n)])
       (for [i (iterate add2 t)
             :let [itmp (sqr i)]
             :while (<= (- itmp (sqr (min2 i))) n)])
       (apply concat)
       (count)))

(defn lamina [n] (time (+ (laminas n 3) (laminas n 4))))

(defn lamina_2
  [lim]
  (->> tiles
       (for [j (range (add2 i) (add2 limx) 2)
             :let [tiles (- (sqr j) (sqr i))]
             :while (<= tiles lim)])
       (for [i (range 1 (inc limx))])
       (let [limx (/ lim 4)])
       (map count)
       (sum)
       (time)))

(defn lamina3
  [lim]
  (time
   (loop [x 1 y (add2 x) res 0]
     (let [tiles (- (sqr y) (sqr x))]
       (if (> tiles lim)
         (if (> x (/ lim 4))
           res
           (recur (inc x) (+ x 3) res))
         (recur x (add2 y) (inc res)))))))

(defn lamina-group
  [lim]
  (loop [x 1 y (add2 x) res []]
    (let [tiles (- (sqr y) (sqr x))]
      (if (> tiles lim)
        (if (> x (/ lim 4))
          res
          (recur (inc x) (+ x 3) res))
        (recur x (add2 y) (conj res tiles))))))

(defn sol174
  [lim]
  (->> (lamina-group lim)
       frequencies
       (group-by val)
       (map #(vector (key %) (count (val %))))
       (filter #(<= (first %) 10))
       (map second)
       (sum)
       (time)))







